enterprise/wan-replication/simulating-max-idle/src/main/java/BouncingSingleMaxIdleSimulator.java

```diff
@@ -0,0 +1,126 @@
+import com.hazelcast.config.Config;
+import com.hazelcast.config.ListenerConfig;
+import com.hazelcast.map.IMap;
+import com.hazelcast.partition.MigrationListener;
+import com.hazelcast.partition.MigrationState;
+import com.hazelcast.partition.ReplicaMigrationEvent;
+
+public class BouncingSingleMaxIdleSimulator extends SingleMaxIdleSimulator {
+    private static final boolean USE_NORMAL_GETS_DURING_MIGRATION = true;
+    private static volatile boolean useNormalGetsA;
+    private static volatile boolean useNormalGetsB;
+
+    private static volatile boolean migratingA;
+    private static volatile boolean migratingB;
+
+    public static void main(String[] args) {
+        AbstractMaxIdleSimulator simulator = new BouncingSingleMaxIdleSimulator();
+        simulator.simulate();
+    }
+
+    private static class MigrationListenerImplA implements MigrationListener {
+
+        @Override
+        public void migrationStarted(MigrationState migrationState) {
+            if (USE_NORMAL_GETS_DURING_MIGRATION) {
+                useNormalGetsA = true;
+            }
+            migratingA = true;
+            if (loggerA != null) {
+                loggerA.info(">> Migration started " + migrationState);
+            }
+        }
+
+        @Override
+        public void migrationFinished(MigrationState migrationState) {
+            if (USE_NORMAL_GETS_DURING_MIGRATION) {
+                useNormalGetsA = false;
+            }
+            migratingA = false;
+            if (loggerA != null) {
+                loggerA.info(">> Migration finished " + migrationState);
+            }
+        }
+
+        @Override
+        public void replicaMigrationCompleted(ReplicaMigrationEvent replicaMigrationEvent) {
+
+        }
+
+        @Override
+        public void replicaMigrationFailed(ReplicaMigrationEvent replicaMigrationEvent) {
+
+        }
+    }
+
+    private static class MigrationListenerImplB implements MigrationListener {
+
+        @Override
+        public void migrationStarted(MigrationState migrationState) {
```

Comment:
I think 2 problems here:

- I believe this listener is called asynchronously. So we can have migration-started event but in reality migration can be completed already. There is a migration interceptor to aware from migration synchronously but it is internal. To see the true affect of migration, migration should be slow. Otherwise it is hard to detect possible latencies reliably in a test.
- migrationStarted is a global event and it has no partition id in it, maybe number of parallel partition can be set to 1 to catch latency effect better.